File: srcr/src/days/day8.rs
use itertools::Itertools;
use std::collections::{HashMap, HashSet};

pub fn solve_day_8(input: &str) {
    let processed_input = input.trim().replace("#", ".");
    let lines: Vec<&str> = processed_input.lines().collect();
    let grid: Vec<Vec<char>> = lines.iter().map(|line| line.chars().collect()).collect();

    let mut char_map: HashMap<char, Vec<(usize, usize)>> = HashMap::new();

    for (row, line) in lines.iter().enumerate() {
        for (col, char) in line.chars().enumerate() {
            if char != '.' {
                char_map.entry(char).or_insert_with(Vec::new).push((row, col));
            }
        }
    }

    let mut anti_nodes: HashSet<(isize, isize)> = HashSet::new();

    for (_char, antennes) in char_map.iter() {
        if antennes.len() > 1 {
            anti_nodes.extend(antennes.iter().map(|&(y, x)| (y as isize, x as isize)));
        }

        for (&(y1, x1), &(y2, x2)) in antennes.iter().tuple_combinations() {
            let dx = (x2 as isize) - (x1 as isize);
            let dy = (y2 as isize) - (y1 as isize);

            let mut factor = 1;
            loop {
                let scaled_dx = dx * factor;
                let scaled_dy = dy * factor;

                let n1 = ((y1 as isize) - scaled_dy, (x1 as isize) - scaled_dx);
                let n2 = ((y2 as isize) + scaled_dy, (x2 as isize) + scaled_dx);

                let in_bounds = |(y, x): (isize, isize)| {
                    y >= 0 && y < grid.len() as isize && x >= 0 && x < grid[0].len() as isize
                };

                if !in_bounds(n1) && !in_bounds(n2) {
                    break;
                }

                if in_bounds(n1) && !antennes.contains(&(n1.0 as usize, n1.1 as usize)) {
                    anti_nodes.insert(n1);
                }

                if in_bounds(n2) && !antennes.contains(&(n2.0 as usize, n2.1 as usize)) {
                    anti_nodes.insert(n2);
                }

                factor += 1;
            }
        }
    }
    println!("Nombre de anti-nodes : {}", anti_nodes.len());
}

File: src/2024/day8.ts
import combinations from "../utils/iterTools";

export function solve_day_8(data: string): string {
    const grid: string[][] = data.split("\n").map((r: string) => r.trim().split(""));

    type ChartMap = {
        [key: string]: [number, number][];
    };

    const isInGrid = (node: [number, number]): boolean => {
        const [y, x] = node;
        return y >= 0 && y < grid.length && x >= 0 && x < grid[0].length;
    };

    const chartMap: ChartMap = {};
    let antiNodes = new Set<string>();

    for (let row = 0; row < grid.length; row++) {
        for (let col = 0; col < grid[row].length; col++) {
            if (grid[row][col] !== ".") {
                const chart: string = grid[row][col];
                if (!chartMap[chart]) {
                    chartMap[chart] = [];
                }
                chartMap[chart].push([row, col]);
            }
        }
    }

    for (const [_, antennes] of Object.entries(chartMap)) {
        if (antennes.length > 1) {
            antennes.forEach(([y, x]) => {
                antiNodes.add(`${y},${x}`);
            });
            }

        for (const [a1, a2] of combinations(antennes, 2)) {
            const [y1, x1] = a1;
            const [y2, x2] = a2;
            const dy: number = y2 - y1;
            const dx: number = x2 - x1;

            let factor: number = 1;

            while (true) {
                const scaled_dx: number = dx * factor;
                const scaled_dy: number = dy * factor;
                const n1: [number, number] = [y1 - scaled_dy, x1 - scaled_dx];
                const n2: [number, number] = [y2 + scaled_dy, x2 + scaled_dx];

                if (!isInGrid(n1) && !isInGrid(n2)) break;

                if (isInGrid(n1)) antiNodes.add(`${n1[0]},${n1[1]}`);
                if (isInGrid(n2)) antiNodes.add(`${n2[0]},${n2[1]}`);

                factor++;
            }
        }
    }
    console.log(antiNodes.size);
    return antiNodes.size.toString();
}